### DISCUSSION TITLE
Header test raises exception

### DISCUSSION THREAD

**Discussion by egisxxegis** [fastapi/fastapi#9037] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Header, testclient, status

app = FastAPI()

@app.get("/")
async def hello(name: str = Header()):
    return {"Hello": name}

client = testclient.TestClient(app, raise_server_exceptions=False)

# succeeds
r = client.get("/", headers={"name": "Tom"})
assert r.status_code == status.HTTP_200_OK

# raises. instead expected to get Response
r = client.get("/", headers={"name": None})
assert r.status_code == status.HTTP_422_UNPROCESSABLE_ENTITY
```

### Description

Test raises exception even though previously it used to return Response object.
- in fastapi 0.85.0 it return Response object - works
- in fastapi 0.87.0 it raises exception
- in fastapi 0.92.0 (newest) it raises exception

part of stack trace:
```python
File ...\.venv\lib\site-packages\httpx\_utils.py", line 55, in normalize_header_value
    return value.encode(encoding or "ascii")
AttributeError: 'NoneType' object has no attribute 'encode'
```

### Operating System

Windows

### Operating System Details

Windows 11

### FastAPI Version

0.92.0

### Python Version

3.10.8

### Additional Context

FastAPI is cool

**Comment by ebottos94** [comment, marked as answer]
Hi @egisxxegis, your second test is wrong, if the header is not preset it will be empty not None type : 

```python
from fastapi import FastAPI, Header, testclient, status

app = FastAPI()

@app.get("/")
async def hello(name: str = Header()):
    return {"Hello": name}

client = testclient.TestClient(app, raise_server_exceptions=False)

# succeeds
r = client.get("/", headers={"name": "Tom"})
assert r.status_code == status.HTTP_200_OK

# raises. instead expected to get Response
r = client.get("/", headers={})
assert r.status_code == status.HTTP_422_UNPROCESSABLE_ENTITY
```

**Comment by egisxxegis** [comment]
Hi @ebottos94, thank You for answering the question - it made me wonder why I no longer can supply `headers={"name": None}` and expect it to be passed as `headers={}`. After digging deeper, I found out that in FastAPI 0.87.0 testclient has been changed to use `httpx` instead of `requests` which brings such subtle difference.
To conclude, `requests` internally converted `{"name": None}` to `{}` while `httpx` expects You (the user) to do the conversion.
Thank You, @ebottos94

**Reply by ebottos94** [reply]
Great! I'm happy to have been helpful.
